File: src/types/task.ts
export type TaskStatus =
  | "pending"
  | "spec_generating"
  | "spec_complete"
  | "ifc_generating"
  | "ifc_complete"
  | "rendering"
  | "completed"
  | "failed"
  | "validating"
  | "quality_check"
  | "approved"
  | "rejected"
  | "revision_in_progress"
  | "revision_complete";

export type QualityStatus = "pending" | "in_progress" | "passed" | "failed" | "conditional";

export type RevisionStatus = "pending" | "in_progress" | "completed" | "cancelled";

export interface AgentResult {
  agent_name: string;
  status: TaskStatus;
  started_at: string;
  completed_at?: string;
  error_message?: string;
  output_summary?: Record<string, any>;
  quality_score?: number;
  warnings: string[];
}

export interface RevisionRecord {
  revision_id: string;
  revision_number: number;
  status: RevisionStatus;
  triggered_by: string;
  reason: string;
  affected_agents: string[];
  phase: string;
  started_at?: string;
  completed_at?: string;
  changes_made: Record<string, any>[];
  quality_check_result?: QualityStatus;
}

export interface ValidationIssue {
  issue_id: string;
  severity: "critical" | "major" | "minor" | "info";
  agent_name: string;
  category: string;
  description: string;
  location?: string;
  recommended_action?: string;
  resolved: boolean;
  resolved_at?: string;
  resolution_notes?: string;
}

export interface WorkflowStatus {
  task_id: string;
  status: string;
  progress: number;
  completed_agents: string[];
  failed_agents: string[];
  revision_number: number;
  phase_1_complete?: boolean;
  current_phase?: string;
  agent_statuses?: Record<string, { status: string; progress: number; current_action?: string }>;
}

export interface Task {
  id: string;
  prompt: string;
  status: TaskStatus;
  progress: number;
  result?: {
    render_path?: string;
    ifc_path?: string;
    thumbnail_path?: string;
  };
  error_message?: string;
  created_at?: string;
  updated_at?: string;
  completed_at?: string;
  quality_score?: number;
  quality_status?: QualityStatus;
  validation_issues?: ValidationIssue[];
  agent_results?: AgentResult[];
  revision_history?: RevisionRecord[];
  current_revision?: RevisionRecord;
  revision_number?: number;
  max_revisions?: number;
  retry_count?: number;
  max_retries?: number;
  context?: Record<string, any>;
  workflow_status?: WorkflowStatus;
}

export interface GalleryItem {
  id: string;
  thumbnail: string;
  image: string;
  ifc: string;
  status: TaskStatus;
  created_at: string;
}

File: service/app/models/task.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from enum import Enum
from datetime import datetime


class GenerateRequest(BaseModel):
    """Request model for generating building renders."""
    prompt: str
    max_revisions: int = 3
    quality: str = "high"  # high, medium, low
    priority: str = "normal"  # low, normal, high
    budget_limit: Optional[float] = None


class TaskStatus(str, Enum):
    """Task status enumeration."""
    PENDING = "pending"
    SPEC_GENERATING = "spec_generating"
    SPEC_COMPLETE = "spec_complete"
    VALIDATING = "validating"
    IFC_GENERATING = "ifc_generating"
    IFC_COMPLETE = "ifc_complete"
    RENDERING = "rendering"
    REVISION_IN_PROGRESS = "revision_in_progress"
    REVISION_COMPLETE = "revision_complete"
    QUALITY_CHECK = "quality_check"
    APPROVED = "approved"
    REJECTED = "rejected"
    COMPLETED = "completed"
    FAILED = "failed"


class RevisionStatus(str, Enum):
    """Revision status enumeration."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class QualityStatus(str, Enum):
    """Quality check status enumeration."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    PASSED = "passed"
    FAILED = "failed"
    CONDITIONAL = "conditional"


class AgentResult(BaseModel):
    """Result from a single agent execution."""
    agent_name: str
    status: TaskStatus
    started_at: datetime
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    output_summary: Optional[Dict[str, Any]] = None
    quality_score: Optional[float] = Field(None, ge=0, le=100)
    warnings: List[str] = []
    progress: int = 0  # Progress within this agent's execution
    current_phase: Optional[str] = None  # Current sub-phase being executed


class RevisionRecord(BaseModel):
    """Record of a revision cycle."""
    revision_id: str
    revision_number: int
    status: RevisionStatus = RevisionStatus.PENDING
    triggered_by: str  # Which agent or system triggered this revision
    reason: str
    affected_agents: List[str] = []
    phase: str = "design"  # "design" or "render"
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    changes_made: List[Dict[str, Any]] = []
    quality_check_result: Optional[QualityStatus] = None


class ValidationIssue(BaseModel):
    """A validation issue found during quality checks."""
    issue_id: str
    severity: str = Field(..., pattern="^(critical|major|minor|info)$")
    agent_name: str
    category: str
    description: str
    location: Optional[str] = None
    recommended_action: Optional[str] = None
    resolved: bool = False
    resolved_at: Optional[datetime] = None
    resolution_notes: Optional[str] = None


class Task(BaseModel):
    """Task model for building generation workflow."""
    id: str
    prompt: str
    status: TaskStatus = TaskStatus.PENDING
    progress: int = 0
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    result: Optional[Dict[str, Any]] = None
    quality_score: float = 0.0
    quality_status: QualityStatus = QualityStatus.PENDING
    validation_issues: List[ValidationIssue] = []
    agent_results: List[AgentResult] = []
    revision_history: List[RevisionRecord] = []
    current_revision: Optional[RevisionRecord] = None
    revision_number: int = 0
    max_revisions: int = 3
    retry_count: int = 0
    max_retries: int = 3
    context: Optional[Dict[str, Any]] = None
    metadata: Optional[Dict[str, Any]] = None  # For structured specs and other metadata
